iarc_roombafilter/scripts/f_ros.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python2
+
+"""
+ROS Interface to the Multi-Target Matching UKF Filter.
+Configuration files should be modified under f_config.py
+
+Takes RoombaObservation and filters them into pose.
+
+TODO(yoonyoungcho) : expose dynamic_reconfigure
+TODO(yoonyoungcho) : convert dt
+TODO(yoonyoungcho) : verify fovcenter ~= Drone Position?
+TODO(yoonyoungcho) : is model-based predictive simulation necessary?
+TODO(yoonyoungcho) : resolve iterative prediction vs. sparse updates
+"""
+
+import rospy
+
+from std_msgs.msg import Header
+from geometry_msgs.msg import Pose, PoseStamped, PoseWithCovariance, PoseWithCovarianceStamped
+from geometry_msgs.msg import Point, Quaternion
+
+from iarc_main.msg import Roomba, RoombaSighting, RoombaList
+import tf
+
+# Filter-Related ...
+from f_utils import *
+import f_config as cfg
+import numpy as np
+import cv2
+
+from f_manager import UKFManager
+
+class UKFManagerROS(object):
+    def __init__(self, dt):
+
+        # unroll parameters ...
+        self._max_targets = rospy.get_param('~max_targets', default=30)
+        self._sigma = np.copy(rospy.get_param('~sigma', default=list(cfg.SIGMAS)))
+        Particle.set_sigmas(np.copy(self._sigma))
+        self._p_keep = rospy.get_param('~p_keep', default=cfg.P_KEEP)
+        self._p_match = rospy.get_param('~p_match', default=cfg.P_MATCH)
+        self._p_clear = rospy.get_param('~p_clear', default=cfg.P_CLEAR)
+        self._obs_topic = rospy.get_param('~obs_topic', default='visible_roombas')
+        self._out_topic = rospy.get_param('~out_topic', default='filtered_roombas')
+        self._rate = rospy.get_param('~rate', default=50) #50hz
+
+        # create ROS interfaces
+        self._tf = tf.TransformListener(True, cache_time=rospy.Duration(5))
+        self._sub = rospy.Subscriber(self._obs_topic, RoombaSighting, self.obs_cb, queue_size=1)
+        self._pub = rospy.Publisher(self._out_topic, RoombaList, queue_size=10)
+        self._mgr = UKFManager(dt, self._sigma)
+        self._t = rospy.Time.now()
+
+        self._obs = None
+
+    def obs_cb(self, msg):
+        """
+        Save observation for processing ...
+        """
+        self._obs = msg
+
+    def obs_proc(self):
+        """
+        Perform filtering according to observation.
+        Decoupled from obs_cb to only process latest observation.
+        """
+        msg = self._obs
+        self._obs = None
+        if msg == None:
+            return
+
+        try:
+            # TODO : kind-of ugly hack to pull stamp into most recent available tf
+            # better way?
+            msg.fov_center.header.stamp = rospy.Time(0)
+            pos = self._tf.transformPoint('map', msg.fov_center)
```

Comment:
This concerns me...

Reply:
concerns me as well. These hacks should be fixed based on test results, etc.